reactive/etcd.py

```diff
@@ -431,6 +432,87 @@ def passive_dismiss_context(cluster):
     cluster.dismiss()
 
 
+@hook('data-storage-attached')
+def format_and_mount_storage():
+    ''' This allows users to request persistent volumes from the cloud provider
+    for the purposes of disaster recovery. '''
+
+    # Query juju for the information about the block storage
+    device_info = storage_get()
+    block = device_info['location']
+
+    if volume_is_mounted(block):
+        hookenv.log('Device is already attached to the system.')
+        hookenv.log('Cowardly refusing to take action against {}'.format(block))
+
+    # Format the device in non-interactive mode
+    cmd = ['mkfs.ext4', device_info['location'], '-F']
+    hookenv.log('Creating filesystem on {}'.format(device_info['location']))
+    hookenv.log('With command: {}'.format(' '.join(cmd)))
+    check_call(cmd)
+
+    # Only attempt migration if directory exists
+    if os.path.isdir('/var/lib/etcd/default'):
+        hookenv.log('Detected existing data, migrating to new location.')
+        # Migrate any existing data
+        cmd = ['mkdir', '-p', '/mnt/etcd-migrate']
+        hookenv.log('With command: {}'.format(' '.join(cmd)))
+        check_call(cmd)
+
+        mount_volume(block, '/mnt/etcd-migrate')
+
+        cmd = ['rsync', '-avz', '/var/lib/etcd/default', '/mnt/etcd-migrate/']
```

Comment:
`v` is excessive and `p` (partials) would be a good flag to add. If rsync is interrupted it'll pick up partials faster